### PR TITLE
KAFKA-4117: Stream partitionassignro cleanup

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Create a new `ClientMetadata` to collapse `Set<String> consumerMemberIds`, `ClientState<TaskId> state`, and `HostInfo hostInfo`.
2. Stop reusing `stateChangelogTopicToTaskIds` and `internalSourceTopicToTaskIds` to access the (sub-)topology's internal repartition and changelog topics for clarity; also use the source topics num.partitions to set the num.partitions for repartition topics, and clarify to NOT have cycles since otherwise the while loop will fail.
3. `ensure-copartition` at the end to modify the number of partitions for repartition topics if necessary to be equal to other co-partition topics.
4. Refactor `ClientState` as well and update the logic of `TaskAssignor` for clarity as well.
5. Change default `clientId` from `applicationId-suffix` to `applicationId-processId` where `processId` is an UUID to avoid conflicts of clientIds that are from different JVMs, and hence conflicts in metrics.
6. Enforce `assignment` partitions to have the same size, and hence 1-1 mapping to `activeTask` taskIds.
7. Remove the `AssignmentSupplier` class by always construct the `partitionsByHostState` before assigning tasks to consumers within a client.
8. Remove all unnecessary member variables in `StreamPartitionAssignor`.
9. Some other minor fixes on unit tests, e.g. remove `test only` functions with java field reflection.

### CODE REVIEW

**Comment by guozhangwang** [comment]
Ping @enothereska @dguy @mjsax for reviews.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 1]
```diff
@@ -57,16 +56,12 @@
...
-    private Map<HostInfo, Set<TopicPartition>> partitionsByHostState;
-    private Cluster metadataWithInternalTopics;
-
 
     private static class AssignedPartition implements Comparable<AssignedPartition> {
         public final TaskId taskId;
         public final TopicPartition partition;
 
-        public AssignedPartition(TaskId taskId, TopicPartition partition) {
+        AssignedPartition(TaskId taskId, TopicPartition partition) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Nit: Can we add `final` to all method parameters (also for all other methods) ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 2]
```diff
@@ -77,6 +72,59 @@ public int compareTo(AssignedPartition that) {
...
+        final HostInfo hostInfo;
+        final Set<String> consumers;
+        ClientState<TaskId> state;
+
+        ClientMetadata(String endPoint) {
+
+            // get the host info if possible
+            if (endPoint != null) {
+                final String[] hostPort = endPoint.split(":");
+                hostInfo = new HostInfo(hostPort[0], Integer.valueOf(hostPort[1]));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Need to check if `hostPost` has exactly two fields to avoid `NullPointerException` and make sure `endPoint` has valid format (more than one `:` would be invalid, too).
Should we catch`NumberFormatException` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 3]
```diff
@@ -77,6 +72,59 @@ public int compareTo(AssignedPartition that) {
...
 
+    private static class ClientMetadata {
+        final HostInfo hostInfo;
+        final Set<String> consumers;
+        ClientState<TaskId> state;
+
+        ClientMetadata(String endPoint) {
+
+            // get the host info if possible
+            if (endPoint != null) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:85]: Can you clarify, why `endPoint` is optional?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 4]
```diff
@@ -187,225 +235,252 @@ public Subscription subscription(Set<String> topics) {
...
+                sourceTopicsByGroup, metadataWithInternalTopics);
+
+        // check if all partitions are assigned, and there are no duplicates of partitions in multiple tasks
+        Set<TopicPartition> allAssingedPartitions = new HashSet<>();
+        Map<Integer, Set<TaskId>> tasksByTopicGroup = new HashMap<>();
+        for (Map.Entry<TaskId, Set<TopicPartition>> entry : partitionsForTask.entrySet()) {
+            Set<TopicPartition> partitions = entry.getValue();
+            for (TopicPartition partition : partitions) {
+                if (allAssingedPartitions.contains(partition)) {
+                    log.warn("stream-thread [{}] Partition {} is assigned to more than one tasks: {}", streamThread.getName(), partition, partitionsForTask);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:397]: Why only warn here? Is this not a severe error?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 5]
```diff
@@ -187,225 +235,252 @@ public Subscription subscription(Set<String> topics) {
...
+                ids = new HashSet<>();
+                tasksByTopicGroup.put(id.topicGroupId, ids);
             }
+            ids.add(id);
+        }
+        for (String topic : allSourceTopics) {
+            for (PartitionInfo partitionInfo : metadataWithInternalTopics.partitionsForTopic(topic)) {
+                TopicPartition partition = new TopicPartition(partitionInfo.topic(), partitionInfo.partition());
+                if (!allAssingedPartitions.contains(partition)) {
+                    log.warn("stream-thread [{}] Partition {} is not assigned to any tasks: {}", streamThread.getName(), partition, partitionsForTask);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:414]: Similar to above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 6]
```diff
@@ -187,225 +235,252 @@ public Subscription subscription(Set<String> topics) {
...
+
+            for (TaskId id : state.assignedTasks) {
+                for (TopicPartition partition : partitionsForTask.get(id)) {
+                    topicPartitions.add(partition);
+                }
+            }
+
+            HostInfo hostInfo = entry.getValue().hostInfo;
+
+            if (hostInfo != null) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Should we move this to the beginning of the loop? If we do not add anything to `partitionsByHostState` there is no reason to create `topicPartitions` in the first place. Or do I miss anything?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 7]
```diff
@@ -558,104 +592,141 @@ public void onAssignment(Assignment assignment) {
...
+            List<String> missingTopics = new ArrayList<>();
+            for (String topic : topicPartitions.keySet()) {
+                List<PartitionInfo> partitions = streamThread.restoreConsumer.partitionsFor(topic);
+                if (partitions == null) {
+                    missingTopics.add(topic);
+                }
+            }
+
+            if (!missingTopics.isEmpty()) {
+                log.warn("stream-thread [{}] Topic {} do not exists but couldn't created as the config '{}' isn't supplied",
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:623]: Is a warning sufficient? Should we throw instead?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 8]
```diff
@@ -558,104 +592,141 @@ public void onAssignment(Assignment assignment) {
...
-    }
+    /**
+     * Internal helper function that creates a Kafka topic
+     *
+     * @param topicPartitions Map that contains the topic names to be created with the number of partitions
+     */
+    private void prepareTopic(Map<String, InternalTopicMetadata> topicPartitions) {
+        // if ZK is specified, prepare the internal source topic before calling partition grouper
+        if (internalTopicManager != null) {
+            log.debug("stream-thread [{}] Starting to validate internal topics in partition assignor.", streamThread.getName());
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Should this log be before the `if` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 9]
```diff
@@ -558,104 +592,141 @@ public void onAssignment(Assignment assignment) {
...
+                internalTopicManager.makeReady(topic, numPartitions);
+
+                // wait until the topic metadata has been propagated to all brokers
+                List<PartitionInfo> partitions;
+                do {
+                    partitions = streamThread.restoreConsumer.partitionsFor(topic.name());
+                } while (partitions == null || partitions.size() != numPartitions);
+            }
+
+            log.info("stream-thread [{}] Completed validating internal topics in partition assignor", streamThread.getName());
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Should this log be after the else-block ?

**Review by mjsax — COMMENTED** [review]
I don't claim to get every detail here. Some suggested improvement though.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:85]: Users may choose not to provide this config in `StreamsConfig`, if they do not need the IQ feature.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java (hunk 2 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: It is checked in line 189 before encoding and propagating to members.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java (hunk 5 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:414]: Ditto above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java (hunk 4 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:397]: User customizable `PartitionGrouper` may choose to do so for their scenarios. I.e. they may want to have some partitions to be assigned to multiple tasks, for broadcasting; or some partitions to not assigned at all, for scaling purposes.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java (hunk 7 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:623]: If the changelog topic does not exist, the restoration will be skipped but not necessarily throwing an error.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 10]
```diff
@@ -57,26 +56,75 @@
...
+    private static class ClientMetadata {
+        final HostInfo hostInfo;
+        final Set<String> consumers;
+        final ClientState<TaskId> state;
+
+        ClientMetadata(final String endPoint) {
+
+            // get the host info if possible
+            if (endPoint != null) {
+                final String[] hostPort = endPoint.split(":");
```
Comment (xvrl) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: this won't work with ipv6 addresses, I think there are some helper methods for this is org.apache.kafka.common.utils.Utils

File: clients/src/main/java/org/apache/kafka/common/Cluster.java [hunk 11]
```diff
@@ -207,6 +207,16 @@ public PartitionInfo partition(TopicPartition topicPartition) {
     }
 
     /**
+     * Get the number of partitions for the given topic
+     * @param topic The topic to get the number of partitions for
+     * @return The number of partitions or null if there is no corresponding metadata
+     */
+    public Integer partitionCountForTopic(String topic) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java:214]: Was there a point to moving this?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java (hunk 10 above)
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Does this need to work with ipv6? Might be worth comparing with `ClientUtils.parseAndValidateAddresses`.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java [hunk 12]
```diff
@@ -66,35 +68,41 @@ public String get(Integer key) {
...
         context.setTime(1);
         written.put(0, "zero");
         changeLogger.add(0);
         written.put(1, "one");
         changeLogger.add(1);
         written.put(2, "two");
         changeLogger.add(2);
-        assertEquals(3, changeLogger.numDirty());
-        assertEquals(0, changeLogger.numRemoved());
+        assertEquals(3, ((Set<Integer>) dirtyField.get(changeLogger)).size());
```
Comment (hachikuji) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java]: Is it really worthwhile to use reflection to avoid exposing a couple trivial getters? If you don't want those methods public, maybe you could keep them package access?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 13]
```diff
@@ -187,225 +235,252 @@ public Subscription subscription(Set<String> topics) {
...
-            state.prevAssignedTasks.addAll(info.standbyTasks);
-            state.capacity = state.capacity + 1d;
+            // add the consumer to the client
+            clientMetadata.addConsumer(consumerId, info);
         }
 
+        log.info("stream-thread [{}] Constructed client metadata {} from the member subscriptions.", streamThread.getName(), clientsMetadata);
 
-        this.topicGroups = streamThread.builder.topicGroups();
+        // ---------------- Step Zero ---------------- //
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:291]: Since this is a fairly complex assignment process, I wonder if it would help to break it down into smaller functions (maybe one for each step?). Otherwise, this is going to be a pretty intimidating chunk of code for newcomers.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 14]
```diff
@@ -448,100 +522,60 @@ public Subscription subscription(Set<String> topics) {
...
+        this.activeTasks = new HashMap<>();
+
+        // the number of assigned partitions should be the same as number of active tasks, which
+        // could be duplicated if one task has more than one assigned partitions
+        if (partitions.size() != info.activeTasks.size()) {
+            TaskAssignmentException ex = new TaskAssignmentException(
+                    String.format("stream-thread [%s] Number of assigned partitions %d is not equal to the number of active taskIds %d" +
+                            ", assignmentInfo=%s", streamThread.getName(), partitions.size(), info.activeTasks.size(), info.toString())
+            );
+            log.error(ex.getMessage(), ex);
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Since this exception is raised, is it necessary to log it here?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 15]
```diff
@@ -558,104 +592,140 @@ public void onAssignment(Assignment assignment) {
...
         }
 
+        // if all topics for this copartition group is repartition topics,
+        // then set the number of partitions to be the maximum of the number of partitions.
         if (numPartitions == -1) {
-            for (InternalTopicConfig topic : internalTopics.values()) {
-                if (copartitionGroup.contains(topic.name())) {
-                    Integer partitions = metadata.partitionCountForTopic(topic.name());
-                    if (partitions != null && partitions > numPartitions) {
+            for (String topic : allRepartitionTopicsNumPartitions.keySet()) {
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: nit: maybe iterate over `entrySet()` instead.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 16]
```diff
@@ -558,104 +592,140 @@ public void onAssignment(Assignment assignment) {
...
             }
         }
-        // enforce co-partitioning restrictions to internal topics reusing internalSourceTopicToTaskIds
-        for (InternalTopicConfig topic : internalTopics.values()) {
-            if (copartitionGroup.contains(topic.name())) {
-                internalSourceTopicToTaskIds
-                    .put(topic, Collections.singleton(new TaskId(-1, numPartitions)));
+
+        // enforce co-partitioning restrictions to repartition topics by updating their number of partitions
+        for (String topic : allRepartitionTopicsNumPartitions.keySet()) {
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Ditto here: use `entrySet()`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignor.java [hunk 17]
```diff
@@ -63,36 +58,38 @@
...
         Set<T> existingTasks = new HashSet<>();
         for (Map.Entry<C, ClientState<T>> entry : states.entrySet()) {
-            this.states.put(entry.getKey(), entry.getValue().copy());
             Set<T> oldTasks = entry.getValue().prevAssignedTasks;
+
             // make sure the previous assignment is balanced
             prevAssignmentBalanced = prevAssignmentBalanced &&
                 oldTasks.size() < 2 * avgNumTasks && oldTasks.size() > avgNumTasks / 2;
+
+            // make sure there is no duplicates
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignor.java]: typo: is -> are

**Review by hachikuji — COMMENTED** [review]
@guozhangwang Added some superficial comments. It might be nice to see if there are any self-contained pieces that can be pulled out of `assign`, which is pretty hard to follow as it stands.

File: clients/src/main/java/org/apache/kafka/common/Cluster.java (hunk 11 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/common/Cluster.java:214]: It's just my evil fun of moving it closer to `partitionsForTopic`..

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:291]: I tried to break it into pieced but with not absolute good solution, mainly because otherwise I would need to create a lot of class member variables to keep track of intermediate results between different stages. I hope the inlined comments can help a bit but admittedly this is quite complex logic that is tricky to explain to newcomers.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java (hunk 12 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java]: For some of these functions, I'd like to not keep the variables as well and only re-compute them when requested in unit tests; and in some other cases they are called from outside of the package as well.

**Comment by guozhangwang** [comment]
Comments addressed.

**Comment by enothereska** [comment]
Ok, this is a fairly big change, could we also report on 1) whether system tests pass and 2) whether the Interactive Queries demos still run. I can help with either or both, just let me know.

**Comment by enothereska** [comment]
The failure seems to be the recurring one from resetting, not related to this.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java (hunk 13 above)
Comment (hachikuji) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:291]: Ack. I think the comments are definitely an improvement. I was just expressing some of my own bewilderment. 😄

**Comment by guozhangwang** [comment]
thanks @enothereska , as for the IQ demos are you referring to the confluentinc/examples? I did make a quick run on those and do not see any failures.

**Comment by enothereska** [comment]
@guozhangwang yes. Then just the streams smoke test and bounce test is left.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java (hunk 12 above)
Comment (hachikuji) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java]: I'm not sure I follow. If you want to remove the variables, then how does accessing them though reflection get you closer? Also, it seems like the fields are still directly accessible from here.

**Comment by guozhangwang** [comment]
I'd prefer to observe the jenkins builds after the PR is merged and if there are errors we can file follow-up PRs. Generally we only pre-check system tests for new features / architectural changes and I feel this PR is none of these, given it is a bit long.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 18]
```diff
@@ -46,37 +46,92 @@
...
+
+        ClientMetadata(final String endPoint) {
+
+            // get the host info if possible
+            if (endPoint != null) {
+                final String host = getHost(endPoint);
+                final Integer port = getPort(endPoint);
+
+                if (host == null || port == null)
+                    throw new ConfigException(String.format("Error parsing host address %s. This should not happen.", endPoint));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: "This should not happen" does not seems to be useful information to the user -- there is an exception -- of course this should not happen.

What about something like: "expected format host:port" or similar?

**Review by mjsax — APPROVED** [review]
One minor comment.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java (hunk 12 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java]: That is a good point, and I think my motivation only stands for public classes like `TopologyBuilder`, whereas all these classes are internal ones.

I can revert these changes if you have strong opinions.

Comment (hachikuji) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java]: No strong opinion, but I'd probably just change the test to access the fields directly rather than through reflection (since they are visible from here).  Otherwise it's just weird.

**Comment by guozhangwang** [comment]
@mjsax @hachikuji comments addressed.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java [hunk 19]
```diff
@@ -20,8 +20,10 @@
 package org.apache.kafka.streams.state.internals;
 
 
+import java.lang.reflect.Field;
```
Comment (hachikuji) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java]: Unused import?

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/StoreChangeLoggerTest.java]: AAAAck..

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 20]
```diff
@@ -187,225 +242,252 @@ public Subscription subscription(Set<String> topics) {
...
-            for (InternalTopicConfig topic : stateChangelogTopics.values()) {
-                Set<TaskId> tasks = stateChangelogTopicToTaskIds.get(topic);
-                if (tasks == null) {
-                    tasks = new HashSet<>();
-                    stateChangelogTopicToTaskIds.put(topic, tasks);
+        Map<TaskId, Set<TopicPartition>> partitionsForTask = streamThread.partitionGrouper.partitionGroups(
+                sourceTopicsByGroup, metadataWithInternalTopics);
+
+        // check if all partitions are assigned, and there are no duplicates of partitions in multiple tasks
+        Set<TopicPartition> allAssingedPartitions = new HashSet<>();
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: typo: allAssingedPartitions

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 21]
```diff
@@ -187,225 +242,252 @@ public Subscription subscription(Set<String> topics) {
...
+        partitionsByHostState = new HashMap<>();
+        for (Map.Entry<UUID, ClientMetadata> entry : clientsMetadata.entrySet()) {
+            HostInfo hostInfo = entry.getValue().hostInfo;
+
+            if (hostInfo != null) {
+                final Set<TopicPartition> topicPartitions = new HashSet<>();
+                final ClientState<TaskId> state = entry.getValue().state;
+
+                for (TaskId id : state.activeTasks) {
+                    for (TopicPartition partition : partitionsForTask.get(id)) {
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Seems you can replace this with this:

``` java
topicPartitions.addAll(partitionsForTask.get(id));
```

Same below.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 22]
```diff
@@ -187,225 +242,252 @@ public Subscription subscription(Set<String> topics) {
...
+        // construct the global partition assignment per host map
+        partitionsByHostState = new HashMap<>();
+        for (Map.Entry<UUID, ClientMetadata> entry : clientsMetadata.entrySet()) {
+            HostInfo hostInfo = entry.getValue().hostInfo;
+
+            if (hostInfo != null) {
+                final Set<TopicPartition> topicPartitions = new HashSet<>();
+                final ClientState<TaskId> state = entry.getValue().state;
+
+                for (TaskId id : state.activeTasks) {
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: One other nit: this loop and the next one (iterating over `state.assignedTasks`) are doing the same thing. Maybe turn it into a function and save some space?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 23]
```diff
@@ -187,225 +242,252 @@ public Subscription subscription(Set<String> topics) {
...
+                        if (numPartitions == -1)
+                            numPartitionsNeeded = true;
+                        else
+                            repartitionTopicMetadata.get(topicName).numPartitions = numPartitions;
                     }
                 }
             }
+        } while (numPartitionsNeeded);
+
+        // augment the metadata with the newly computer number of partitions for all the
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: typo: computer -> computed

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 24]
```diff
@@ -187,225 +242,252 @@ public Subscription subscription(Set<String> topics) {
...
+        // repartition source topics
+        Map<TopicPartition, PartitionInfo> allRepartitionTopicPartitions = new HashMap<>();
+        for (Map.Entry<String, InternalTopicMetadata> entry : repartitionTopicMetadata.entrySet()) {
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:352]: Seems this could be a function as well. For example:

``` java
Map<TopicPartition, PartitionInfo> partitions(Map<String, InternalTopicMetadata>);
```

(I'm looking for small independent chunks of code that can be taken out of this function.)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 25]
```diff
@@ -187,225 +242,252 @@ public Subscription subscription(Set<String> topics) {
...
-        internalSourceTopicToTaskIds = new HashMap<>();
-        Map<Integer, Set<String>> sourceTopicGroups = new HashMap<>();
-        Map<Integer, Collection<InternalTopicConfig>> internalSourceTopicGroups = new HashMap<>();
+        // parse the topology to determine the repartition source topics,
+        // making sure they are created with the number of partitions as
+        // the maximum of the depending sub-topologies source topics' number of partitions
+        Map<Integer, TopologyBuilder.TopicsInfo> topicGroups = streamThread.builder.topicGroups();
+
+        Map<String, InternalTopicMetadata> repartitionTopicMetadata = new HashMap<>();
         for (Map.Entry<Integer, TopologyBuilder.TopicsInfo> entry : topicGroups.entrySet()) {
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: This looks like another case where extracting a function could help reduce some of the noise. Also, seems you can just iterate over the values now?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 26]
```diff
@@ -187,225 +242,252 @@ public Subscription subscription(Set<String> topics) {
...
+            for (InternalTopicConfig topic: entry.getValue().repartitionSourceTopics.values()) {
+                repartitionTopicMetadata.put(topic.name(), new InternalTopicMetadata(topic));
+            }
         }
 
+        boolean numPartitionsNeeded;
+        do {
+            numPartitionsNeeded = false;
+
+            for (Map.Entry<Integer, TopologyBuilder.TopicsInfo> entry : topicGroups.entrySet()) {
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Ditto here: seems we don't need the key? Same for the nested loop over `topicGroups`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 27]
```diff
@@ -187,225 +242,252 @@ public Subscription subscription(Set<String> topics) {
...
+                        // if we still have not find the right number of partitions,
+                        // another iteration is needed
+                        if (numPartitions == -1)
+                            numPartitionsNeeded = true;
+                        else
+                            repartitionTopicMetadata.get(topicName).numPartitions = numPartitions;
                     }
                 }
             }
+        } while (numPartitionsNeeded);
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Is `numPartitionsNeeded` equivalent to `numPartitions == -1`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java [hunk 28]
```diff
@@ -52,13 +54,15 @@ private ClientState(Set<T> activeTasks, Set<T> assignedTasks, Set<T> prevActiveT
     }
 
     public ClientState<T> copy() {
-        return new ClientState<>(new HashSet<>(activeTasks), new HashSet<>(assignedTasks),
+        return new ClientState<>(new HashSet<>(activeTasks), new HashSet<T>(standbyTasks), new HashSet<>(assignedTasks),
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java]: nit: the `T` parameter is unnecessary.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 29]
```diff
@@ -187,225 +242,252 @@ public Subscription subscription(Set<String> topics) {
...
-            Set<String> consumers = entry.getValue();
-            ClientState<TaskId> state = states.get(processId);
+        // within the client, distribute tasks to its owned consumers
+        Map<String, Assignment> assignment = new HashMap<>();
+        for (Map.Entry<UUID, ClientMetadata> entry : clientsMetadata.entrySet()) {
+            Set<String> consumers = entry.getValue().consumers;
+            ClientState<TaskId> state = entry.getValue().state;
 
             ArrayList<TaskId> taskIds = new ArrayList<>(state.assignedTasks.size());
             final int numActiveTasks = state.activeTasks.size();
```
Comment (hachikuji) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:485]: Are the two loops below equivalent to this?

``` java
taskIds.addAll(state.activeTasks);
taskIds.addAll(state.standbyTasks);
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java (hunk 22 above)
Comment (hachikuji) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: On second thought, it seems `activeTasks` are a subset of `assignedTasks`? Then is the first loop even needed?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: Ack. We need the order for taskIds, which is an ArrayList; but we don't need it for the set of topic-partitions.

**Comment by guozhangwang** [comment]
@hachikuji Comments addressed; about extracting a couple for loops into separate functions: personally I'm a bit inclined not to do it if it is only called once, and the procedure itself do not have significant semantic meaning except construct some data structure from other structures.

**Comment by hachikuji** [comment]
@guozhangwang Fair enough. I just noticed there was a lot of boilerplate conversion in this code between different objects and maps. It adds a lot of noise to the overflow flow.

**Comment by hachikuji** [comment]
LGTM

**Comment by guozhangwang** [comment]
Merged to trunk.
